### PR TITLE
Standalone Lite: Strategy base classes and registry

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Adds the Strategy base classes. Implementations and the corresponding tests will follow in the other PRs.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

### CODE REVIEW

File: src/lightning_lite/strategies/strategy.py [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+    def teardown(self) -> None:
+        """This method is called to teardown the training process.
+
+        It is the right place to release memory and free other resources.
+        """
+        self.precision_plugin.teardown()
+        assert self.accelerator is not None
+        self.accelerator.teardown()
+        self.checkpoint_io.teardown()
+
```
Comment (awaelchli) [inline src/lightning_lite/strategies/strategy.py:297]: Reviewer: For Lite, there is no teardown that happens automatically. I would suggest to either
A) Remove it
B) Provide a method specific to freeing memory (that the user could call via lite.free_memory()) for example.

File: src/lightning_lite/strategies/strategy.py [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+            checkpoint: dict containing model and trainer state
+            filepath: write-target file's path
+            storage_options: parameter for how to save to storage, passed to ``CheckpointIO`` plugin
+        """
+        if self.is_global_zero:
+            self.checkpoint_io.save_checkpoint(checkpoint, filepath, storage_options=storage_options)
+
+    def get_module_state_dict(self, module: Module) -> Dict[str, Union[Any, Tensor]]:
+        """Returns model state."""
+        return module.state_dict()
```
Comment (awaelchli) [inline src/lightning_lite/strategies/strategy.py:253]: Reviewer note: Some of these hooks now take the module as input, as opposed to the PL implementation which would reference self.lightning_module normally

File: src/lightning_lite/strategies/strategy.py [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+        self._precision_plugin = precision_plugin
+
+    def _configure_launcher(self) -> None:
+        """Attach the launcher based on Strategy."""
+
+    def setup_environment(self) -> None:
+        """Setup any processes or distributed connections.
+
+        This is called before the LightningModule/DataModule setup hook which allows the user to access the accelerator
+        environment before setup is complete.
```
Comment (awaelchli) [inline src/lightning_lite/strategies/strategy.py:102]: TODO: update this text

File: src/lightning_lite/strategies/strategy.py (hunk 1 above)
Comment (Borda) [reply src/lightning_lite/strategies/strategy.py:295]: I would call for C) but dot have any strong preference here :otter:

**Review by Borda — APPROVED** [review]

File: src/lightning_lite/strategies/strategy.py (hunk 1 above)
Comment (justusschock) [reply src/lightning_lite/strategies/strategy.py:295]: Can't we call it automatically after `run` is finished?

Comment (awaelchli) [reply src/lightning_lite/strategies/strategy.py:295]: @justusschock We could but I guess what I'm trying to say here is that teardown() is a broad term and in PL it does a lot of things, while in Lite it will never be able to conceptually do the same as in PL. Some example: Moving the LightningModule to CPU, unwrapping the module's sync-batchnorm layers, etc. For this reason, I was hoping in Lite we could make the APIs much more focused, and then in PL we could still keep the very general purpose teardown.

File: src/lightning_lite/strategies/parallel.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    _get_process_group_backend_from_env,
+    all_gather_ddp_if_available,
+    get_default_process_group_backend_for_device,
+    ReduceOp,
+)
+from lightning_lite.utilities.rank_zero import rank_zero_deprecation
+
+
+class ParallelStrategy(Strategy, ABC):
+    """Plugin for training with multiple processes in parallel."""
```
Comment (kaushikb11) [inline src/lightning_lite/strategies/parallel.py:38]:
```suggestion
    """Strategy for training with multiple processes in parallel."""
```

Comment (awaelchli) [reply src/lightning_lite/strategies/parallel.py:38]: After all this time, these old plugin references still pop up 😄

File: src/lightning_lite/strategies/strategy.py (hunk 1 above)
Comment (justusschock) [reply src/lightning_lite/strategies/strategy.py:295]: In PL we have setup and teardown, so in Lite we could have setup_device and teardown_device, which only does the device-related stuff and could be called automatically. This could then be extended in PL to do the more general purpose things

Comment (awaelchli) [reply src/lightning_lite/strategies/strategy.py:295]: Yes exactly, that's what I mean by focused functions. So we are aligned. 

Lite doesn't do this today, which is why I'm in favor of A/B until we will actually add it to Lite's workflow. As we are currently focused on just moving things to the package in this first iteration.

I'm creating an issue for it so we can work on it.

File: src/lightning_lite/strategies/parallel.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    def distributed_sampler_kwargs(self) -> Dict[str, Any]:
+        distributed_sampler_kwargs = dict(
+            num_replicas=len(self.parallel_devices) if self.parallel_devices is not None else 0, rank=self.global_rank
+        )
+        return distributed_sampler_kwargs
+
+    @property
+    def torch_distributed_backend(self) -> str:
+        """Deprecated property."""
+        rank_zero_deprecation(
```
Comment (carmocca) [inline src/lightning_lite/strategies/parallel.py:95]: Why not keep the deprecations in a thin subclass in PL?

File: src/lightning_lite/strategies/registry.py [hunk 6]
```diff
@@ -112,11 +111,8 @@ def __str__(self) -> str:
         return "Registered Strategies: {}".format(", ".join(self.keys()))
 
 
-StrategyRegistry = _StrategyRegistry()
-
-
-def call_register_strategies(base_module: str) -> None:
+def call_register_strategies(registry: _StrategyRegistry, base_module: str) -> None:
```
Comment (carmocca) [inline src/lightning_lite/strategies/registry.py:114]: This is a breaking change. But I don't think this should have been exposed anyways. So if you are going to add an argument, I would also prefix it

File: src/lightning_lite/strategies/strategy.py [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+        Returns: Model parallel context.
+        """
+        yield
+
+    def pre_backward(self, tensor: Tensor, module: Optional[Module]) -> None:
+        """Run before precision plugin executes backward."""
+
+    def backward(self, tensor: Tensor, module: Optional[Module], *args: Any, **kwargs: Any) -> None:
+        r"""Forwards backward-calls to the precision plugin."""
+        self.pre_backward(tensor, module)
```
Comment (carmocca) [inline src/lightning_lite/strategies/strategy.py:167]: I thought you had no use for pre and post in Lite?

**Review by carmocca — APPROVED** [review]

File: src/lightning_lite/strategies/parallel.py (hunk 5 above)
Comment (awaelchli) [reply src/lightning_lite/strategies/parallel.py:95]: Oh, I forgot to remove this. This can safely be removed in Lite and PL will remove theirs also for 1.8.

File: src/lightning_lite/strategies/strategy.py (hunk 7 above)
Comment (awaelchli) [reply src/lightning_lite/strategies/strategy.py:167]: Hmm, there are some strategies that implement, but it's not relevant for Lite atm. 
It's probably better to remove it for now yes.

File: src/lightning_lite/strategies/registry.py (hunk 6 above)
Comment (carmocca) [reply src/lightning_lite/strategies/registry.py]: Yes

**Comment by Borda** [comment]
@kaushikb11 ^^ :otter:

**Review by justusschock — APPROVED** [review]
